File: src/Configuration/ConfigFileHelpers.cs
using System;
using System.IO;

/// <summary>
/// Provides methods for determining configuration file locations.
/// </summary>
public static class ConfigFileHelpers
{
    public static string? GetConfigFileName(ConfigFileScope scope)
    {
        var path = FindConfigFile(scope);
        if (path != null) return path;

        return GetYamlConfigFileName(scope);
    }

    public static string? FindConfigFile(ConfigFileScope scope, bool forceCreate = false)
    {
        var yamlPath = GetYamlConfigFileName(scope);
        if (File.Exists(yamlPath))
        {
            ConsoleHelpers.WriteDebugLine($"Found YAML config file at: {yamlPath}");
            return yamlPath;
        }

        var iniPath = GetIniConfigFileName(scope);
        if (File.Exists(iniPath))
        {
            ConsoleHelpers.WriteDebugLine($"Found INI config file at: {iniPath}");
            return iniPath;
        }

        if (forceCreate && !string.IsNullOrEmpty(yamlPath))
        {
            FileHelpers.WriteAllText(yamlPath!, string.Empty);
            return yamlPath;
        }

        return null;
    }
    
    public static string? GetScopeDirectoryPath(ConfigFileScope scope)
    {
        return scope switch
        {
            ConfigFileScope.Global => GetGlobalScopeDirectory(),
            ConfigFileScope.User => GetUserScopeDirectory(),
            ConfigFileScope.Local => GetLocalScopeDirectory(),
            _ => null
        };
    }
    
    public static string? GetLocationDisplayName(ConfigFileScope scope)
    {
        var configPath = FindConfigFile(scope);
        if (configPath != null)
        {
            return $"{configPath} ({scope.ToString().ToLower()})";
        }
        
        var directory = GetScopeDirectoryPath(scope);
        var badDirectory = string.IsNullOrEmpty(directory);
        if (badDirectory) return null;

        var defaultPath = Path.Combine(directory!, YAML_CONFIG_NAME);
        return $"{defaultPath} ({scope.ToString().ToLower()})";
    }
    
    public static string GetGlobalScopeDirectory()
    {
        var parent = GetGlobalScopeParentDirectory();
        var configDir = Path.Combine(parent, CONFIG_DIR_NAME);
        ConsoleHelpers.WriteDebugLine($"Global config directory: {configDir}");
        return configDir;
    }

    public static string GetUserScopeDirectory()
    {
        var parent = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        return Path.Combine(parent, CONFIG_DIR_NAME);
    }

    public static string GetLocalScopeDirectory()
    {
        var existingYamlFile = FileHelpers.FindFileSearchParents(CONFIG_DIR_NAME, YAML_CONFIG_NAME);
        if (existingYamlFile != null) return Path.GetDirectoryName(existingYamlFile)!;

        var existingIniFile = FileHelpers.FindFileSearchParents(CONFIG_DIR_NAME, INI_CONFIG_NAME);
        if (existingIniFile != null) return Path.GetDirectoryName(existingIniFile)!;

        return Path.Combine(Directory.GetCurrentDirectory(), CONFIG_DIR_NAME);
    }

    private static string GetGlobalScopeParentDirectory()
    {
        var isLinuxOrMac = Environment.OSVersion.Platform == PlatformID.Unix || Environment.OSVersion.Platform == PlatformID.MacOSX;
        if (isLinuxOrMac)
        {
            var xdgDataHome = Environment.GetEnvironmentVariable("XDG_DATA_HOME");
            if (!string.IsNullOrEmpty(xdgDataHome))
            {
                ConsoleHelpers.WriteDebugLine($"On Linux/Mac... using XDG_DATA_HOME: {xdgDataHome}");
                return xdgDataHome;
            }
            else
            {
                var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                var homeLocalShareDir = Path.Combine(homeDir, ".local", "share");
                ConsoleHelpers.WriteDebugLine($"On Linux/Mac... using HOME/.local/share: {homeLocalShareDir}");
                return homeLocalShareDir;
            }
        }

        var commonAppData = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData);
        ConsoleHelpers.WriteDebugLine($"On Windows... using CommonApplicationData: {commonAppData}");
        return commonAppData;
    }

    private static string? GetYamlConfigFileName(ConfigFileScope scope)
    {
        var path = GetScopeDirectoryPath(scope);
        return path != null
            ? Path.Combine(path, YAML_CONFIG_NAME)
            : null;
    }

    private static string? GetIniConfigFileName(ConfigFileScope scope)
    {
        var path = GetScopeDirectoryPath(scope);
        return path != null
            ? Path.Combine(path, INI_CONFIG_NAME)
            : null;
    }

    private const string CONFIG_DIR_NAME = $".{Program.Name}";
    private const string YAML_CONFIG_NAME = "config.yaml";
    private const string INI_CONFIG_NAME = "config";
}

File: src/commonjs/Configuration/ConfigFileHelpers.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { ConfigFileScope } from './ConfigFileScope';

/**
 * Provides methods for determining configuration file locations.
 */
export class ConfigFileHelpers {
    public static GetConfigFileName(scope: ConfigFileScope): string | null {
        const foundPath = ConfigFileHelpers.FindConfigFile(scope);
        if (foundPath != null) return foundPath;

        return ConfigFileHelpers.GetYamlConfigFileName(scope);
    }

    public static FindConfigFile(scope: ConfigFileScope, forceCreate: boolean = false): string | null {
        const yamlPath = ConfigFileHelpers.GetYamlConfigFileName(scope);
        if (yamlPath && fs.existsSync(yamlPath)) {
            console.debug(`Found YAML config file at: ${yamlPath}`);
            return yamlPath;
        }

        const iniPath = ConfigFileHelpers.GetIniConfigFileName(scope);
        if (iniPath && fs.existsSync(iniPath)) {
            console.debug(`Found INI config file at: ${iniPath}`);
            return iniPath;
        }

        if (forceCreate && yamlPath) {
            fs.writeFileSync(yamlPath, '');
            return yamlPath;
        }

        return null;
    }
    
    public static GetScopeDirectoryPath(scope: ConfigFileScope): string | null {
        switch (scope) {
            case ConfigFileScope.Global:
                return ConfigFileHelpers.GetGlobalScopeDirectory();
            case ConfigFileScope.User:
                return ConfigFileHelpers.GetUserScopeDirectory();
            case ConfigFileScope.Local:
                return ConfigFileHelpers.GetLocalScopeDirectory();
            default:
                return null;
        }
    }
    
    public static GetLocationDisplayName(scope: ConfigFileScope): string | null {
        const configPath = ConfigFileHelpers.FindConfigFile(scope);
        if (configPath != null) {
            return `${configPath} (${ConfigFileScope[scope].toLowerCase()})`;
        }
        
        const directory = ConfigFileHelpers.GetScopeDirectoryPath(scope);
        if (!directory) return null;

        const defaultPath = path.join(directory, ConfigFileHelpers.YAML_CONFIG_NAME);
        return `${defaultPath} (${ConfigFileScope[scope].toLowerCase()})`;
    }
    
    public static GetGlobalScopeDirectory(): string {
        const parent = ConfigFileHelpers.GetGlobalScopeParentDirectory();
        const configDir = path.join(parent, ConfigFileHelpers.CONFIG_DIR_NAME);
        console.debug(`Global config directory: ${configDir}`);
        return configDir;
    }

    public static GetUserScopeDirectory(): string {
        const parent = os.homedir();
        return path.join(parent, ConfigFileHelpers.CONFIG_DIR_NAME);
    }

    public static GetLocalScopeDirectory(): string {
        const existingYamlFile = ConfigFileHelpers.FindFileSearchParents(ConfigFileHelpers.CONFIG_DIR_NAME, ConfigFileHelpers.YAML_CONFIG_NAME);
        if (existingYamlFile != null) return path.dirname(existingYamlFile);

        const existingIniFile = ConfigFileHelpers.FindFileSearchParents(ConfigFileHelpers.CONFIG_DIR_NAME, ConfigFileHelpers.INI_CONFIG_NAME);
        if (existingIniFile != null) return path.dirname(existingIniFile);

        return path.join(process.cwd(), ConfigFileHelpers.CONFIG_DIR_NAME);
    }

    private static GetGlobalScopeParentDirectory(): string {
        const isLinuxOrMac = process.platform === 'linux' || process.platform === 'darwin';
        if (isLinuxOrMac) {
            const xdgDataHome = process.env.XDG_DATA_HOME;
            if (xdgDataHome) {
                console.debug(`On Linux/Mac... using XDG_DATA_HOME: ${xdgDataHome}`);
                return xdgDataHome;
            } else {
                const homeDir = os.homedir();
                const homeLocalShareDir = path.join(homeDir, '.local', 'share');
                console.debug(`On Linux/Mac... using HOME/.local/share: ${homeLocalShareDir}`);
                return homeLocalShareDir;
            }
        }

        // Windows - use ALLUSERSPROFILE or fallback
        const commonAppData = process.env.ALLUSERSPROFILE || 'C:\\ProgramData';
        console.debug(`On Windows... using CommonApplicationData: ${commonAppData}`);
        return commonAppData;
    }

    private static GetYamlConfigFileName(scope: ConfigFileScope): string | null {
        const dirPath = ConfigFileHelpers.GetScopeDirectoryPath(scope);
        return dirPath != null
            ? path.join(dirPath, ConfigFileHelpers.YAML_CONFIG_NAME)
            : null;
    }

    private static GetIniConfigFileName(scope: ConfigFileScope): string | null {
        const dirPath = ConfigFileHelpers.GetScopeDirectoryPath(scope);
        return dirPath != null
            ? path.join(dirPath, ConfigFileHelpers.INI_CONFIG_NAME)
            : null;
    }

    private static readonly CONFIG_DIR_NAME = '.cycod'; // ProgramInfo.ConfigDirName equivalent
    private static readonly YAML_CONFIG_NAME = 'config.yaml';
    private static readonly INI_CONFIG_NAME = 'config';

    private static FindFileSearchParents(dirName: string, fileName: string): string | null {
        let currentDir = process.cwd();
        
        while (currentDir !== path.dirname(currentDir)) {
            const configDir = path.join(currentDir, dirName);
            const configFile = path.join(configDir, fileName);
            
            if (fs.existsSync(configFile)) {
                return configFile;
            }
            
            currentDir = path.dirname(currentDir);
        }
        
        return null;
    }
}